### PR TITLE
feat: Modal v4

### PR DESCRIPTION
**Pull request by suevalov**
# Purpose of PR

Extract Modal components to a separate package.

* Buttons are aligned to right and there's no possibility to align them to left.
* Added ability to specify `initialFocusRef` when the Modal is opened.
* Added more documentation about components.
* No other changes besides that.

### CODE REVIEW

File: packages/components/modal/src/Modal.mdx [hunk 1]
```diff
@@ -42,12 +53,20 @@ function ModalExample() {
             <Modal.Header title="Title" />
             <Modal.Content>Hello from controlled modal window</Modal.Content>
             <Modal.Controls>
-              <Button buttonType="positive" onClick={() => setShown(false)}>
-                Confirm
-              </Button>
-              <Button buttonType="muted" onClick={() => setShown(false)}>
+              <Button
+                size="small"
+                variant="transparent"
```
Comment (gui-santos) [inline packages/components/modal/src/Modal.mdx:58]:
```suggestion
                variant="secondary"
```

File: packages/components/modal/src/Modal.mdx [hunk 2]
```diff
@@ -92,12 +111,12 @@ function ModalExample() {
             <Modal.Header title="Title" />
             <Modal.Content>Hello from controlled modal window</Modal.Content>
             <Modal.Controls>
-              <Button buttonType="positive" onClick={() => setShown(false)}>
-                Confirm
-              </Button>
-              <Button buttonType="muted" onClick={() => setShown(false)}>
+              <Button variant="transparent" onClick={() => setShown(false)}>
```
Comment (gui-santos) [inline packages/components/modal/src/Modal.mdx:114]:
```suggestion
              <Button variant="secondary" onClick={() => setShown(false)}>
```

File: packages/components/modal/src/Modal.mdx [hunk 3]
```diff
@@ -142,12 +161,12 @@ function ModalExample() {
...
```
Comment (gui-santos) [inline packages/components/modal/src/Modal.mdx:164]:
```suggestion
              <Button variant="secondary" onClick={() => setShown(false)}>
```

File: packages/components/modal/src/ModalConfirm/ModalConfirm.tsx [hunk 4]
```diff
@@ -47,7 +47,12 @@ export interface ModalConfirmProps {
   /**
    * The intent of the ModalConfirm. Used for the secondary Button.
    */
-  secondaryIntent?: 'primary' | 'positive' | 'negative' | 'transparent';
+  secondaryIntent?:
+    | 'primary'
+    | 'positive'
+    | 'negative'
+    | 'transparent'
+    | 'secondary';
```
Comment (gui-santos) [inline packages/components/modal/src/ModalConfirm/ModalConfirm.tsx:55]: I thought, based on the designs, that the secondary button would always be `secondary `

File: packages/components/modal/src/ModalConfirm/ModalConfirm.tsx [hunk 5]
```diff
@@ -174,28 +191,19 @@ export function ModalConfirm({
...
-              ) : (
-                <Fragment>
-                  {confirmButton}
-                  {secondaryButton}
-                  {cancelButton}
-                </Fragment>
-              )}
+              {cancelButton}
+              {secondaryButton}
+              {confirmButton}
```
Comment (gui-santos) [inline packages/components/modal/src/ModalConfirm/ModalConfirm.tsx:204]: Will we still support three buttons in the modal?

File: packages/components/modal/src/ModalHeader/ModalHeader.styles.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+export function getModalHeaderStyles() {
+  return {
+    root: css({
+      padding: `${tokens.spacingM} ${tokens.spacingM} ${tokens.spacingM} ${tokens.spacingL}`,
+      borderRadius: `${tokens.borderRadiusMedium} ${tokens.borderRadiusMedium} 0 0`,
+      borderBottom: `1px solid ${tokens.gray300}`,
+    }),
+    heading: css({
+      flexGrow: 1,
+    }),
```
Comment (gui-santos) [inline packages/components/modal/src/ModalHeader/ModalHeader.styles.ts:13]: I think we could eliminate this class if use `justifyContent="space-between"` in the `Flex` component (the one using the root class)

File: packages/components/modal/src/ModalHeader/ModalHeader.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      testId={testId}
+      alignItems="center"
+      flexShrink={0}
+    >
+      <Subheading
+        as="h1"
+        isTruncated
+        marginBottom="none"
+        className={styles.heading}
+      >
```
Comment (gui-santos) [inline packages/components/modal/src/ModalHeader/ModalHeader.tsx:39]:
```suggestion
    <Flex
      {...otherProps}
      className={cx(styles.root, className)}
      testId={testId}
      alignItems="center"
      justifyContent="space-between"
    >
      <Subheading
        as="h1"
        isTruncated
        marginBottom="none"
      >
```

☝️  That's what I meant in the other comment

File: packages/components/modal/stories/ModalLauncher.stories.tsx [hunk 8]
```diff
@@ -1,12 +1,11 @@
...
 import { Paragraph } from '@contentful/f36-typography';
 
-import { ModalLauncher } from './ModalLauncher';
+import { ModalLauncher } from '../src/ModalLauncher/ModalLauncher';
 import { Button } from '@contentful/f36-button';
-import { Modal } from '../Modal';
+import { Modal } from '../src/Modal';
 
-function DefaultStory() {
+export const Basic = () => {
```
Comment (gui-santos) [inline packages/components/modal/stories/ModalLauncher.stories.tsx:8]: This story is not showing up, and I think it is because it's missing the default export with storybook setup you know?

The one that looks like this:

```js
export default {
  title: 'Utilities/ModalLauncher',
  component: ModalLauncher,
  // ...
} as Meta;
```

File: packages/components/skeleton/src/SkeletonContainer/SkeletonContainer.tsx [hunk 9]
```diff
@@ -48,108 +48,104 @@ export interface SkeletonContainerProps extends CommonProps {
...
-    clipId,
-    gradientId,
-    animateId,
-    ...otherProps
-  }) => {
-    const uniqueClipId = useId(clipId, 'cf-ui-skeleton-clip');
-    const uniqueGradientId = useId(gradientId, 'cf-ui-skeleton-clip-gradient');
-    const randomAnimateId = useId(undefined, 'animation');
-    const uniqueAnimateId = animateId || randomAnimateId;
+export function SkeletonContainer({
```
Comment (gui-santos) [inline packages/components/skeleton/src/SkeletonContainer/SkeletonContainer.tsx:51]: Shouldn't this be in another PR? Or is it relevant to Modal changes?

File: packages/forma-36-react-components/src/components/Dropdown/README.mdx [hunk 10]
```diff
@@ -38,7 +38,7 @@ function DropdownExample() {
       toggleElement={
         <Button
           size="small"
-          buttonType="muted"
+          variant="transparent"
```
Comment (gui-santos) [inline packages/forma-36-react-components/src/components/Dropdown/README.mdx:41]:
```suggestion
          variant="secondary"
```

I think the variant `muted` is now called `secondary`

The transparent one is the old "naked" button and if I remember correctly we wanted to recommend the use of secondary buttons instead of the naked ones

Am I remembering this correctly or dreaming @fabe and @domarku?

maybe I'm wrong, but my problem with transparent buttons is that they don't differentiate themselves from the rest of the UI enough

Screenshot 2021-08-12 at 17 51 52

File: packages/forma-36-website/src/components/ComponentSource.js [hunk 11]
```diff
@@ -60,7 +60,11 @@ class ComponentSource extends React.Component {
                 <LiveEditor className={styles.editor} />
               </React.Fragment>
             )}
-            <Button isFullWidth buttonType="muted" onClick={this.handleToggle}>
+            <Button
+              isFullWidth
+              variant="transparent"
```
Comment (gui-santos) [inline packages/forma-36-website/src/components/ComponentSource.js:65]:
```suggestion
              variant="secondary"
```

File: packages/forma-36-website/src/components/DocFormatter.js [hunk 12]
```diff
@@ -87,9 +88,15 @@ const DocFormatter = ({ frontmatter, dataFromReadme, children }) => {
               <>
                 <Button
                   size="small"
-                  buttonType={activeSection === 'design' ? 'primary' : 'muted'}
+                  variant={
+                    activeSection === 'design' ? 'primary' : 'transparent'
```
Comment (gui-santos) [inline packages/forma-36-website/src/components/DocFormatter.js:92]:
```suggestion
                    activeSection === 'design' ? 'primary' : 'secondary'
```

File: packages/forma-36-website/src/components/DocFormatter.js [hunk 13]
```diff
@@ -98,8 +105,14 @@ const DocFormatter = ({ frontmatter, dataFromReadme, children }) => {
                 <Button
                   size="small"
                   className={styles.buttonRight}
-                  buttonType={activeSection === 'develop' ? 'primary' : 'muted'}
-                  icon={activeSection === 'develop' && 'CheckCircleTrimmed'}
+                  variant={
+                    activeSection === 'develop' ? 'primary' : 'transparent'
```
Comment (gui-santos) [inline packages/forma-36-website/src/components/DocFormatter.js:109]:
```suggestion
                    activeSection === 'develop' ? 'primary' : 'secondary'
```

**Review by gui-santos — CHANGES_REQUESTED** [review]
This is looking good

I just have add a bunch of questions about using `secondary` button instead of `transparent`
some CSS suggestion and questions

And one general comment is that i feel that this PR has more stuff than just changes in the Modal, right? There are changes in other things that don't seem to be related to Modal (e.g.: SkeletonContainer)
I think it would be easier to review if split the PRs
One PR per task basically

**Comment by gui-santos** [comment]
I removed my comments about the `transparent` because there are many of them

But my general comment is that every time we see a `variant="muted"` we should convert to `variant="secondary"`

I believe it was decided we are deprecating `naked` or `transparent` (is this right @fabe and @domarku?)

File: packages/forma-36-react-components/src/components/Dropdown/README.mdx (hunk 10 above)
Comment (burakukula) [reply packages/forma-36-react-components/src/components/Dropdown/README.mdx:41]: @gui-santos we decided to re-implement naked version but we called it "transparent" now. It is needed for cases, like `close` IconButton (now it will be handled by button component). So we actually need this variant.

Comment (gui-santos) [reply packages/forma-36-react-components/src/components/Dropdown/README.mdx:41]: ok, but it's being used in places that are not `IconButton` in the PR

Like in the screenshot I posted, I believe those buttons should be `primary` or `secondary` I know it's just documentation, but I think we want to improve the UX of the docs as well 

Screenshot 2021-08-13 at 13 12 16

I understand why the `IconButton` uses transparent variant, but the IconButton feels "clickable" hahahah
I guess we just wanted to use all the Button features in the IconButton without the border, but for normal buttons we fall in the same problem we had with the `naked` variant. When it's used alone, without other buttons, the naked/transparent button is just text that the user needs to guess if it's clickable or not 🤷‍♂️

File: packages/components/modal/src/Modal.styles.ts [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+      overflow: 'hidden',
+      display: 'flex',
+      flexDirection: 'column',
+    }),
+    props.allowHeightOverflow
+      ? css({
+          overflow: 'auto',
+          maxHeight: 'none',
+        })
+      : null,
```
[the remark this answers is not in the document]
Comment (suevalov) [reply packages/components/modal/src/Modal.styles.ts:31]: This property is exposed as a part of UI Extensions SDK, so it can be removed without breaking people apps and extensions.

https://github.com/contentful/ui-extensions-sdk/blob/b0bbaad477114f7929d974b8bce8c3c03f05f4bf/lib/types.ts#L477

File: packages/components/modal/src/Modal.styles.ts [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+  size: ModalSizeType;
+  position: ModalPositionType;
+  allowHeightOverflow?: boolean;
+  className?: string;
+}) {
+  const modal = cx(
+    css({
+      transition: `opacity ${tokens.transitionDurationLong} ${tokens.transitionEasingDefault}, transform ${tokens.transitionDurationDefault} ${tokens.transitionEasingCubicBezier}`,
+      opacity: '0.5',
+      margin: '50px',
```
[the remark this answers is not in the document]
Comment (suevalov) [reply packages/components/modal/src/Modal.styles.ts]: Updated to the closest possible one

File: packages/components/modal/src/ModalConfirm/ModalConfirm.tsx (hunk 5 above)
Comment (suevalov) [reply packages/components/modal/src/ModalConfirm/ModalConfirm.tsx:204]: Yeah, a good question. I'd remove `secondaryButton` completely and make "cancelButton" to look like a secondary

File: packages/components/modal/src/ModalHeader/ModalHeader.styles.ts (hunk 6 above)
Comment (suevalov) [reply packages/components/modal/src/ModalHeader/ModalHeader.styles.ts]: thanks, done

File: packages/components/modal/src/ModalHeader/ModalHeader.tsx (hunk 7 above)
Comment (suevalov) [reply packages/components/modal/src/ModalHeader/ModalHeader.tsx]: done

File: packages/components/skeleton/src/SkeletonContainer/SkeletonContainer.tsx (hunk 9 above)
Comment (suevalov) [reply packages/components/skeleton/src/SkeletonContainer/SkeletonContainer.tsx:51]: I had to fix it in order to procceed.

File: packages/components/modal/stories/ModalLauncher.stories.tsx (hunk 8 above)
Comment (suevalov) [reply packages/components/modal/stories/ModalLauncher.stories.tsx]: thanks for the suggestion, done

File: packages/forma-36-react-components/src/components/Dropdown/README.mdx (hunk 10 above)
Comment (suevalov) [reply packages/forma-36-react-components/src/components/Dropdown/README.mdx:41]: Got it, rolled back to "secondary" here.

**Comment by denkristoffer** [comment]
@gui-santos I don't think we're deprecating `transparent`, we need it to be able to replace the `IconButton` component, don't we?

**Comment by gui-santos** [comment]
> @gui-santos I don't think we're deprecating `transparent`, we need it to be able to replace the `IconButton` component, don't we?

@denkristoffer yes! @burakukula reminded me about the case of the `IconButton` 👌 

now I know, thanks everyone 🙏

File: packages/components/modal/src/ModalConfirm/ModalConfirm.tsx (hunk 5 above)
Comment (suevalov) [reply packages/components/modal/src/ModalConfirm/ModalConfirm.tsx]: Removed secondary button after talking to @fabe and @domarku

File: packages/components/modal/src/types.ts [hunk 16]
```diff
@@ -0,0 +1,10 @@
+export type ModalSizeType =
+  | 'small'
+  | 'medium'
+  | 'large'
+  | 'fullWidth'
+  | 'zen'
+  | string
+  | number;
+
+export type ModalPositionType = 'center' | 'top';
```
[the remark this answers is not in the document]
Comment (suevalov) [reply packages/components/modal/src/types.ts:10]: the same as `allowHeightOverflow` it's exposed to UI Extensions SDK: https://github.com/contentful/ui-extensions-sdk/blob/b0bbaad477114f7929d974b8bce8c3c03f05f4bf/lib/types.ts#L478

**Comment by suevalov** [comment]
> And one general comment is that i feel that this PR has more stuff than just changes in the Modal, right? There are changes in other things that don't seem to be related to Modal (e.g.: SkeletonContainer)
I think it would be easier to review if split the PRs
One PR per task basically

Some of the changes are basically the fixes for the bugs that were found while I was migrating the component.

File: packages/components/modal/src/Modal.tsx [hunk 17]
```diff
@@ -106,6 +126,27 @@ export function Modal({
     topOffset,
   };
```
Comment (gui-santos) [inline packages/components/modal/src/Modal.tsx:127]: Could you fix this in your PR?

I guess in the past we were doing this weird thing where we destructure the props in the argument of the function and then we assemble everything again in a const called props 🤷 

```jsx
function Component({ children, ...otherProps}) {
  // no const props here
  
  return <div {...otherProps}>{children}</div>
}
```

I've seen this pattern with duplication of props in other components and I really don't know where it came from hahahaha

File: packages/components/modal/src/ModalConfirm/ModalConfirm.tsx [hunk 18]
```diff
@@ -15,15 +15,11 @@ export interface ModalConfirmProps {
   /**
    * Function that will be called when the confirm button is clicked. This does not close the ModalConfirm.
    */
-  onConfirm: Function;
-  /**
-   * Function that will be called when the secondary button is clicked. This does not close the ModalConfirm.
-   */
-  onSecondary?: Function;
+  onConfirm(): void;
```
Comment (gui-santos) [inline packages/components/modal/src/ModalConfirm/ModalConfirm.tsx:18]: TIL this is the same as `onConfirm: () => void`

File: packages/components/modal/src/ModalHeader/ModalHeader.tsx [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+  return (
+    <Flex
+      {...otherProps}
+      className={cx(styles.root, className)}
+      testId={testId}
+      alignItems="center"
+      justifyContent="space-between"
+    >
+      <Subheading as="h1" isTruncated marginBottom="none">
```
Comment (gui-santos) [inline packages/components/modal/src/ModalHeader/ModalHeader.tsx:34]: Is this correct?

What happens if the page behind the modal has an h1?

File: packages/forma-36-react-components/src/components/Dropdown/README.mdx (hunk 10 above)
Comment (gui-santos) [reply packages/forma-36-react-components/src/components/Dropdown/README.mdx:41]: You forgot this one

```suggestion
          variant="secondary"
```

**Review by gui-santos — APPROVED** [review]
LGTM

I have only one favor to ask that is the refactoring of the duplicated props
It's something that you didn't introduce, but it would be nice to refactor it every time we touch something 🙏

File: packages/components/modal/src/ModalHeader/ModalHeader.tsx (hunk 19 above)
Comment (suevalov) [reply packages/components/modal/src/ModalHeader/ModalHeader.tsx:34]: I think we apply aria attribute to hide the background when the modal is opened. But it was always `h1`.

Comment (gui-santos) [reply packages/components/modal/src/ModalHeader/ModalHeader.tsx:34]: yeah I'm not the a11y expert, I was just wondering
if it works, that's great

File: packages/components/modal/README.mdx [hunk 20]
```diff
@@ -0,0 +1,3 @@
+# @contentful/f36-modal
+
+This package is part of [Forma 36](https://github.com/contentful/forma-36). See the repo for more details.
```
Comment (burakukula) [inline packages/components/modal/README.mdx:3]: Could we migrate the README file?

File: packages/components/modal/src/Modal.mdx [hunk 21]
```diff
@@ -3,10 +3,13 @@ title: 'Modal'
 type: 'component'
 status: 'stable'
 slug: /components/modal/
-github: 'https://github.com/contentful/forma-36/tree/master/packages/forma-36-react-components/src/components/Modal'
-storybook: 'https://f36-storybook.contentful.com/?path=/story/components-modal--default'
+github: 'https://github.com/contentful/forma-36/tree/master/packages/components/modal/src'
```
Comment (burakukula) [inline packages/components/modal/src/Modal.mdx:6]: ah, sorry, it's here. hmmm, why we have 2 README files?

Comment (suevalov) [reply packages/components/modal/src/Modal.mdx:6]: we have three: `Modal`, `ModalConfirm` and `ModalLauncher`

File: packages/components/modal/src/Modal.tsx [hunk 22]
```diff
@@ -15,30 +18,27 @@ const ModalSizesMapper = {
   zen: '100vw',
```
Comment (burakukula) [inline packages/components/modal/src/Modal.tsx:18]: what a strange name. ☯️ Do we want to keep it?

Comment (suevalov) [reply packages/components/modal/src/Modal.tsx:18]: yes, it's used inside MarkdownEditor

File: packages/components/modal/src/Modal.tsx (hunk 17 above)
Comment (suevalov) [reply packages/components/modal/src/Modal.tsx:127]: I guess this one is needed here, because `props` are passed to a children function

Comment (suevalov) [reply packages/components/modal/src/Modal.tsx:127]: and it's the only place in the modal package where we do this

**Review by burakukula — APPROVED** [review]
